### PR TITLE
Added sigmoid function class in exponential.py

### PR DESCRIPTION
**Pull request by jhanwar**
#### Brief description of what is fixed or changed

Added sigmoid function class

#### Release Notes

- functions
  - Added a class of Sigmoid function Added various methods (fdiff, eval, taylor_term, as_real_imag, _eval_is_real, _eval_is_algebric, _taylor, _eval_rewrite_as_tanh)

### CODE REVIEW

**Comment by jhanwar** [comment]
#### Release Notes
*functions
       *Added a class of Sigmoid function Added various methods  (fdiff, eval, taylor_term, as_real_imag, _eval_is_real, _eval_is_algebric, _taylor, _eval_rewrite_as_tanh)

**Comment by oscarbenjamin** [comment]
I'm not sure it makes sense to define a sigmoid function as a distinct function. It's easily defined in terms of the existing elementary functions (i.e. exp). The module you're adding this to defines `exp`, `log` and `LambertW`. The first two there are pretty fundamental. The `LambertW` function is related to the others but cannot be explicitly defined in terms of them.

What are the advantages of having a sigmoid function rather than just using an expression involving exp?

**Comment by oscargus** [comment]
I would say that it is convenient to have sigmoid defined as such. Will be more readable code etc.

@jhanwar I am pro this functionality, but the code require a bit of a cleanup, documentation, and unit tests. It would also be good if one can change the sharpness of the sigmoid.

One may consider adding a new file of activation functions (for neural networks/machine learning) where maybe this fits better.

File: sympy/functions/elementary/exponential.py [hunk 1]
```diff
@@ -885,3 +885,89 @@ def _eval_is_algebraic(self):
...
+            raise ArgumentIndexError(self, argindex)
+
+    @classmethod
+    def eval(cls, arg):
+
+        if arg is S.Zero:
+            return S.Half
+
+        arg = sympify(arg)
+        print("hello")
```
Comment (oscargus) [inline sympy/functions/elementary/exponential.py]: Delete

**Comment by oscargus** [comment]
`_rewrite_as_exp` would also be natural to add.

**Comment by oscargus** [comment]
Reading up a bit more on https://en.wikipedia.org/wiki/Sigmoid_function it may be better to actually provide a larger class of Sigmoid functions. The one you provided is the standard Logistic function https://en.wikipedia.org/wiki/Logistic_function with a certain set of constants.

On the other hand, while that will be more general it also involves more challenges.

**Comment by oscarbenjamin** [comment]
What is the advantage of having an explicit sigmoid function when we can easily define it with `exp`?

I'm not saying that to oppose this: I just want to understand. Can we do anything special with the knowledge that it is a sigmoid function that would be more difficult using an expression with Eq? Do sigmoid functions have any particular algebraic properties that we can exploit?

There are many different mathematical functions that can be defined from standard elementary functions. I can't see examples of these being explicitly defined as special classes in SymPy though.

**Comment by oscargus** [comment]
As I said: convenience. Why should we be able to use an arbitrary base in log when we simply can write it as `log(x)/log(base)` (which sympy currently does exactly that for) ? Sort of the same answer.

**Comment by oscarbenjamin** [comment]
For convenience we can have a function that returns the corresponding expression. What's the advantage of defining a new class and then reimplementing the methods like diff that the expression would have already had?

**Comment by oscargus** [comment]
Yes, that was what I came up to as well if you read the last part of my first comment of today. The potential benefit though would have been to have special printing etc and maybe some non-obvious rewrites. Still, here printing is a problem as most of the sigmoid functions have some additional parameters that it is not obvious how to print in a coherent way.

**Comment by czgdp1807** [comment]
It looks like that this PR can be closed. Adding a `Could close` label.

**Comment by jhanwar** [comment]
As @oscargus mentioned in above comments that we could have a separate file of neural network activation functions. But I guess, this is not concluded yet.

**Comment by jhanwar** [comment]
As @oscargus mentioned in above comments that we could have a separate file of neural network activation functions. But I guess, this is not concluded yet.

**Comment by czgdp1807** [comment]
IMHO, Adding neural network activation functions won't be of much use in `SymPy` as they relate more numerical tasks rather than symbolic tasks. 
Well, What about other CAS? Do they have a collection of activation functions pre-defined?

**Comment by czgdp1807** [comment]
It appears that Wolfram Alpha supports `sigmoid`, `relu`, etc. For ease we can just define somewhere in a file a list of activation functions.
```python
sigmoid = 1/(1+exp(-x))
...
# other activation functions
```
Users can directly import these functions for ease of use. But I think defining a specialised class will be of little use since SymPy ideally should be capable of performing operations on them by using pre-written code.
I tried the operations given at, https://www.wolframalpha.com/input/?i=sigmoid on `isympy` and they worked out pretty well,
```python
>>> sigmoid = 1/(1 + exp(-x))
>>> sigmoid
   1   
───────
     -x
1 + ℯ  
>>> sigmoid.diff(x)
    -x    
   ℯ      
──────────
         2
⎛     -x⎞ 
⎝1 + ℯ  ⎠ 
>>> series(sigmoid, x)
         3     5        
1   x   x     x     ⎛ 6⎞
─ + ─ - ── + ─── + O⎝x ⎠
2   4   48   480        
>>> integrate(sigmoid, x)
       ⎛     -x⎞
x + log⎝1 + ℯ  ⎠
>>> limit(sigmoid, x, oo)
1
>>> limit(sigmoid, x, -oo)
0
```

**Comment by czgdp1807** [comment]
I will make a follow up PR to add the list of common activation functions under, `sympy/sympy/functions/special`. Let me know of any comments on this idea.

**Comment by czgdp1807** [comment]
Closing this for now. Please let me know if this needs to be re-opened. Thanks for your contributions.

**Pull request closed without merging**